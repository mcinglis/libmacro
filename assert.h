
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_ASSERT_H
#define LIBMACRO_ASSERT_H


#include <assert.h>

#include <libpp/map.h>          // PP_MAP
#include <libpp/separators.h>   // PP_SEP_SEMICOLON


#ifdef NO_ASSERT
    #define ASSERT( ... )
#else

    // @public
    // Takes a variable number of scalar expressions, and `assert`s each
    // one individually. This provides better assertion error reporting than
    // techniques like `assert( e1 && e2 && ... )`.
    #define ASSERT( ... ) \
        PP_MAP( assert, PP_SEP_SEMICOLON, __VA_ARGS__ )

#endif


#endif


