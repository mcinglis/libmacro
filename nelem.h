
// Copyright 2015  Malcolm Inglis <http://minglis.id.au>
//
// This file is part of Libmacro.
//
// Libmacro is free software: you can redistribute it and/or modify it under
// the terms of the GNU Affero General Public License as published by the
// Free Software Foundation, either version 3 of the License, or (at your
// option) any later version.
//
// Libmacro is distributed in the hope that it will be useful, but WITHOUT
// ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
// FITNESS FOR A PARTICULAR PURPOSE. See the GNU Affero General Public
// License for more details.
//
// You should have received a copy of the GNU Affero General Public License
// along with Libmacro. If not, see <https://gnu.org/licenses/>.


#ifndef LIBMACRO_NELEM_H
#define LIBMACRO_NELEM_H


// Gives the number of elements in the array `XS`. Be very careful that
// you only call this with an *array* variable, and not a pointer-to-array.
// Note that this evaluates to a constant expression.
#define NELEM( ... ) \
    ( ( sizeof ( __VA_ARGS__ ) ) / ( sizeof ( ( __VA_ARGS__ )[ 0 ] ) ) )


#endif

