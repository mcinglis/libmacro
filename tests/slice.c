
#include <stdio.h>

#include "../slice.h"
#include "../assert.h"
#include "../nelem.h"


int main( void )
{
    int const xs[] = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };

    printf( "Testing subset slice...\n" );
    int const ws[] = { SLICE( xs, 3, 4 ) };
    ASSERT( NELEM( ws ) == 4,
            ws[ 0 ] == xs[ 3 ],
            ws[ 1 ] == xs[ 4 ],
            ws[ 2 ] == xs[ 5 ],
            ws[ 3 ] == xs[ 6 ] );
    ( void ) ws;

    printf( "Testing total slice...\n" );
    int const ys[] = { SLICE( xs, 0, 6 ) };
    ASSERT( NELEM( ys ) == 6,
            ys[ 0 ] == xs[ 0 ],
            ys[ 1 ] == xs[ 1 ],
            ys[ 2 ] == xs[ 2 ],
            ys[ 3 ] == xs[ 3 ],
            ys[ 4 ] == xs[ 4 ],
            ys[ 5 ] == xs[ 5 ] );
    ( void ) ys;

    printf( "Testing empty slice...\n" );
    int const zs[] = { 0, SLICE( xs, 2, 0 ) };
    ASSERT( NELEM( zs ) == 1 );
    ( void ) zs;

    printf( "SLICE() tests passed.\n" );
}

