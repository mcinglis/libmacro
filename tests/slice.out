Testing subset slice...
Testing total slice...
Testing empty slice...
SLICE() tests passed.
